### ISSUE TITLE
TypeError: undefined is not an object (evaluating 'n.uibuilder')

### ISSUE THREAD

**Issue by devleaks** [TotallyInformation/node-red-contrib-uibuilder#37]
Hello,
I'm trying to build a Vuejs app as UI builder for node. I use the latest version fo vue-cli (3.X).

In node ~/.node-red/uibuilder I create a new vue app, compile it, run it, it works.
I'm trying to model like React app as explained in the wiki. So I'm trying to use the helper functions uibuilder provides. In the HelloWorld component, I simply add the following line:

import uibuilder from './uibuilderfe.js' 

uibuilderfe.js  which I copied from node-red-contrib-uibuilder folder.
The vue app compiles but no longer work, I get the error message in the console:

TypeError: undefined is not an object (evaluating 'n.uibuilder')

I'm not even using the uibuilderfe, just loading it.

Any idea?

**Comment by devleaks** [comment]
I guess uibuilderfe is a wrapper around web sockets.
I'd rather use a native web socket package/interface. What is the connect string (url ws://localhost:1880/...) to get messages from node-red uibuilder?
P.

**Comment by TotallyInformation** [comment]
Hi, the uibuilder fe code self executes so that beginners immediately get access to the object and the event handlers that it initialises (so that you can easily detect when a new msg arrives from Node-RED).

It is possible that this is clashing with something in Vue?

If you turn on the FE debugging using the flag in the node, you may be able to see where things are going wrong.

Unfortunately, I never got round to trying to build an example in Vue as I'd already given up on it after previously trying to build a generic library around it (the forerunner to uibuilder). It was too limiting in the way that it worked - or maybe I was too inexperienced in using it. 

If the uibuilderfe debug doesn't throw up where the error lies, you will need to add some more debugging to Vue I'm afraid. But we can work through it together. Unfortunately, I'll not be much available for a week or so after today but I will try to respond to questions and updates here.


What _does_ work is [MoonJS](https://hackernoon.com/introducing-moon-1d44a99635f0) which is very similar in use to Vue but rather slimmer. Not sure if you are able to switch if you can't get Vue to work.

**Comment by TotallyInformation** [comment]
OK, as a test, I took the MoonJS example in the WIKI, replaced references to Moon with ones for Vue, turned on debug and I can see messages from the fe library going both ways. Need to do some tweaking to the code to get Vue to actually work but it is a hopeful start.

If you are using Chrome, there is a Vue addin that may also help with debugging. See the browser dev console for details.

Example index.js:

```html
<!doctype html>
<html lang="en">
  <!--
    This is the default, template html for uibuilder.
    It is only meant to demonstrate the use of JQuery to dynamically
    update the ui based on incoming/outgoing messages from/to the
    Node-RED server.

    You will want to alter this to suite your own needs. To do so,
    copy this file to <userDir>/uibuilder/<url>/src.
  -->
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=yes">

    <!-- See https://goo.gl/OOhYW5 -->
    <link rel="manifest" href="./manifest.json">
    <meta name="theme-color" content="#3f51b5">

    <!-- Used if adding to homescreen for Chrome on Android. Fallback for manifest.json -->
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="application-name" content="Node-RED UI Builder">

    <!-- Used if adding to homescreen for Safari on iOS -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <meta name="apple-mobile-web-app-title" content="Node-RED UI Builder">

    <!-- Homescreen icons for Apple mobile use if required
        <link rel="apple-touch-icon" href="/images/manifest/icon-48x48.png">
        <link rel="apple-touch-icon" sizes="72x72" href="/images/manifest/icon-72x72.png">
        <link rel="apple-touch-icon" sizes="96x96" href="/images/manifest/icon-96x96.png">
        <link rel="apple-touch-icon" sizes="144x144" href="/images/manifest/icon-144x144.png">
        <link rel="apple-touch-icon" sizes="192x192" href="/images/manifest/icon-192x192.png">
    -->

    <title>Node-RED UI Builder</title>
    <meta name="description" content="Node-RED UI Builder">

    <link rel="icon" href="./images/node-blue.ico">

    <!-- OPTIONAL: Normalize is used to make things the same across browsers. Index is for your styles -->
    <link rel="stylesheet" href="./vendor/normalize.css/normalize.css">
    <link rel="stylesheet" href="./index.css">

</head>
<body>
    <div id="app">
        <h1>
            UIbuilder + Moon for Node-RED
        </h1>
        <p>
            This is a uibuilder test using <a href="http://moonjs.ga/">Moon.JS</a> as a front-end library.
            See the
            <a href="https://github.com/TotallyInformation/node-red-contrib-uibuilder">node-red-contrib-uibuilder</a>
            README for details on how to use UIbuilder.
        </p>
        <h2>Dynamic Data (via Moon)</h2>

        <p>
            Messages: Received={{msgsReceived}}, Sent: {{msgsSent}}
        </p><p>
            Control Messages Received: {{msgsControl}}
        </p>

        <pre m-html="hLastRcvd"></pre>
        <pre m-html="hLastSent"></pre>

        <h2>Simple input using Moon</h2>
        <p><button m-on:click="increment">Increment</button> Click Counter: {{counterBtn}}. Click on the button to increment, it sends the data back to Node-RED as well.</p>

    </div>

    <!-- These MUST be in the right order. Note no leading / -->
    <!-- REQUIRED: Socket.IO is loaded only once for all instances
                   Without this, you don't get a websocket connection -->
    <script src="/uibuilder/socket.io/socket.io.js"></script>
    <!-- Note no leading / -->
    <!-- OPTIONAL: JQuery can be removed if not required -->
    <!--<script src="./vendor/jquery/dist/jquery.min.js"></script>-->

    <script src="vendor/vue/dist/vue.js"></script>

    <!-- REQUIRED: Sets up Socket listeners and the msg object -->
    <script src="./uibuilderfe.min.js"></script>
    <!-- OPTIONAL: You probably want this. Put your custom code here -->
    <script src="./index.js"></script>

</body>
</html>
```

Example index.js:

```javascript
const app1 = new Vue({
    el: "#app",
    data: {
        startMsg    : "Vue has started, waiting for messages",
        feVersion   : '',
        counterBtn  : 0,
        msgsReceived: 0,
        msgsControl : 0,
        msgsSent    : 0,
        msgRecvd    : '[Nothing]',
        msgSent     : '[Nothing]',
        msgCtrl     : '[Nothing]',
        inputText   : ''
    }, // --- End of data --- //
    computed: {
        hLastRcvd: {
            get: function() {
                const msgRecvd = this.get('msgRecvd')
                if (typeof msgRecvd === 'string') return 'Last Message Received = ' + msgRecvd
                else return 'Last Message Received = ' + this.callMethod('syntaxHighlight', [msgRecvd])
            }
        },
        hLastSent: {
            get: function() {
                const msgSent = this.get('msgSent')
                if (typeof msgSent === 'string') return 'Last Message Sent = ' + msgSent
                else return 'Last Message Sent = ' + this.callMethod('syntaxHighlight', [msgSent])
            }
        },
        hMsgCtrl: {
            get: function() {
                const msgCtrl = this.get('msgCtrl')
                if (typeof msgCtrl === 'string') return 'Last Message Sent = ' + msgCtrl
                //else return 'Last Message Sent = ' + this.callMethod('syntaxHighlight', [msgCtrl])
                else return 'Last Message Sent = ' + JSON.stringify(msgCtrl)
            }
        }
    }, // --- End of computed --- //
    methods: {
        increment: function() {
            // Increment the count by one
            this.set('counterBtn', this.get('counterBtn') + 1)
            let topic = (this.get('msgRecvd')).topic || 'uibuilder/moon'
            uibuilder.send( { 'topic': topic, 'payload': { 'type': 'counterBtn', 'btnCount': this.get('counterBtn'), 'message': this.get('inputText') } } )
        },
        // return formatted HTML version of JSON object
        syntaxHighlight: function(json) {
            json = JSON.stringify(json, undefined, 4)
            json = json.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
            return json.replace(/("(\\u[a-zA-Z0-9]{4}|\\[^u]|[^\\"])*"(\s*:)?|\b(true|false|null)\b|-?\d+(?:\.\d*)?(?:[eE][+\-]?\d+)?)/g, function (match) {
                var cls = 'number'
                if (/^"/.test(match)) {
                    if (/:$/.test(match)) {
                        cls = 'key'
                    } else {
                        cls = 'string'
                    }
                } else if (/true|false/.test(match)) {
                    cls = 'boolean'
                } else if (/null/.test(match)) {
                    cls = 'null'
                }
                return '<span class="' + cls + '">' + match + '</span>'
            })
        } // --- End of syntaxHighlight --- //
    }, // --- End of methods --- //
    hooks: { // Available hooks: init,mounted,updated,destroyed

        mounted: function(){
            console.debug('app mounted - setting up uibuilder watchers')

            this.set( 'feVersion', uibuilder.get('version'))

            // If msg changes - msg is updated when a standard msg is received from Node-RED over Socket.IO
            // Note that you can also listen for 'msgsReceived' as they are updated at the same time
            // but newVal relates to the attribute being listened to.
            uibuilder.onChange('msg', function(newVal){
                console.info('property msg changed!', newVal)
                app1.set( 'msgRecvd', newVal)
            })
            // As noted, we could get the msg here too
            uibuilder.onChange('msgsReceived', function(newVal){
                console.info('New msg sent FROM Node-RED over Socket.IO. Total Count: ', newVal)
                app1.set( 'msgsReceived', newVal)
            })


            // If a message is sent back to Node-RED
            uibuilder.onChange('sentMsg', function(newVal){
                console.info('property sentMsg changed!', newVal)
                app1.set( 'msgSent', newVal)
            })
            uibuilder.onChange('msgsSent', function(newVal){
                console.info('New msg sent TO Node-RED over Socket.IO. Total Count: ', newVal)
                app1.set('msgsSent', newVal )
            })


            // If we receive a control message from Node-RED
            uibuilder.onChange('ctrlMsg', function(newVal){
                console.info('property msgCtrl changed!', newVal)
                app1.set( 'msgCtrl', newVal)
            })
            uibuilder.onChange('msgsCtrl', function(newVal){
                console.info('New CONTROL msg sent FROM Node-RED over Socket.IO. Total Count: ', newVal)
                app1.set('msgsControl', newVal )
            })

            // If Socket.IO connects/disconnects
            uibuilder.onChange('ioConnected', function(newVal){
                console.info('Socket.IO Connection Status Changed: ', newVal)
                app1.set('socketConnectedState', newVal )
            })

        } // --- End of mounted hook --- //
    } // --- End of hooks --- //
}) // --- End of app1 --- //

// EOF
```

**Comment by devleaks** [comment]
OK, I got your demo to work as well.

The problem is coming from the way you "import" uibuilderfe. If you use import, or import {} from file  `this` is null in the uibuilderfe script. (I don't understand why yet.) and since you force the context to this (which is null) the statement this.uibuilder fails.

In your demo, you load everything in a single html file and uibuilder gets defined at the top level. You load all the code you need (socket.io) before. That's not the way vue developer work. They handcraft careful isolated components, and I still cannot figure out how to load your function in a single component, and not at the global scope level.

**Comment by devleaks** [comment]
Here is your code in a single file component.
I'd love to load uibuilder here only.
I struggle to load it from your file.
Have a look at the first few lines.
How can I import your script and get your object uibuilder?
```
<script>
// How to import your script in this file only?
import uibuilder from './uibuilderfe.js';         // does not work
// const uibuilder = require('./uibuilderfe.js'); // does not work

export default {
  name: 'UiBuilder',
  props: {
    msg: String
  },
  data: function () {
    return {
      startMsg    : "Vue has started, waiting for messages",
      feVersion   : '',
      counterBtn  : 0,
      msgsReceived: 0,
      msgsControl : 0,
      msgsSent    : 0,
      msgRecvd    : '[Nothing]',
      msgSent     : '[Nothing]',
      msgCtrl     : '[Nothing]',
      inputText   : ''
    }
  }, // --- End of data --- //
  computed: {
      hLastRcvd: function() {
          const msgRecvd = this.msgRecvd
          if (typeof msgRecvd === 'string') return 'Last Message Received = ' + msgRecvd
          else return 'Last Message Received = ' + this.syntaxHighlight(msgRecvd)
      },
      hLastSent: function() {
          const msgSent = this.msgSent
          if (typeof msgSent === 'string') return 'Last Message Sent = ' + msgSent
          else return 'Last Message Sent = ' + this.syntaxHighlight(msgSent)
      },
      hMsgCtrl: function() {
          const msgCtrl = this.msgCtrl
          if (typeof msgCtrl === 'string') return 'Last Message Sent = ' + msgCtrl
          //else return 'Last Message Sent = ' + this.callMethod('syntaxHighlight', [msgCtrl])
          else return 'Last Message Sent = ' + JSON.stringify(msgCtrl)
      }
  }, // --- End of computed --- //
  methods: {
    increment: function() {
        // Increment the count by one
        this.counterBtn = this.counterBtn + 1
        let topic = this.msgRecvd.topic || 'uibuilder/vue'
        uibuilder.send( { 'topic': topic, 'payload': { 'type': 'counterBtn', 'btnCount': this.counterBtn, 'message': this.inputText } } )
    },
    // return formatted HTML version of JSON object
    syntaxHighlight: function(json) {
        json = JSON.stringify(json, undefined, 4)
        json = json.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
        return json.replace(/("(\\u[a-zA-Z0-9]{4}|\\[^u]|[^\\"])*"(\s*:)?|\b(true|false|null)\b|-?\d+(?:\.\d*)?(?:[eE][+\-]?\d+)?)/g, function (match) {
            var cls = 'number'
            if (/^"/.test(match)) {
                if (/:$/.test(match)) {
                    cls = 'key'
                } else {
                    cls = 'string'
                }
            } else if (/true|false/.test(match)) {
                cls = 'boolean'
            } else if (/null/.test(match)) {
                cls = 'null'
            }
            return '<span class="' + cls + '">' + match + '</span>'
        })
      } // --- End of syntaxHighlight --- //
    }, // --- End of methods --- //

    // Available hooks: init,mounted,updated,destroyed
    mounted: function(){
      console.debug('app mounted - setting up uibuilder watchers')

      var vueApp = this

      vueApp.feVersion = uibuilder.get('version')

      // If msg changes - msg is updated when a standard msg is received from Node-RED over Socket.IO
      // Note that you can also listen for 'msgsReceived' as they are updated at the same time
      // but newVal relates to the attribute being listened to.
      uibuilder.onChange('msg', function(newVal){
          console.info('property msg changed!', newVal)
          vueApp.msgRecvd = newVal
      })
      // As noted, we could get the msg here too
      uibuilder.onChange('msgsReceived', function(newVal){
          console.info('New msg sent FROM Node-RED over Socket.IO. Total Count: ', newVal)
          vueApp.msgsReceived = newVal
      })

      // If a message is sent back to Node-RED
      uibuilder.onChange('sentMsg', function(newVal){
          console.info('property sentMsg changed!', newVal)
          vueApp.msgSent = newVal
      })
      uibuilder.onChange('msgsSent', function(newVal){
          console.info('New msg sent TO Node-RED over Socket.IO. Total Count: ', newVal)
          vueApp.msgsSent = newVal
      })

      // If we receive a control message from Node-RED
      uibuilder.onChange('ctrlMsg', function(newVal){
          console.info('property msgCtrl changed!', newVal)
          vueApp.msgCtrl = newVal
      })
      uibuilder.onChange('msgsCtrl', function(newVal){
          console.info('New CONTROL msg sent FROM Node-RED over Socket.IO. Total Count: ', newVal)
          vueApp.msgsControl = newVal
      })

      // If Socket.IO connects/disconnects
      uibuilder.onChange('ioConnected', function(newVal){
          console.info('Socket.IO Connection Status Changed: ', newVal)
          vueApp.socketConnectedState = newVal
      })

  } // --- End of mounted hook --- //
}
</script>

<template>
  <div>
    <h1>
        UIbuilder + Vue.js for Node-RED
    </h1>
    <p>
        This is a uibuilder test using <a href="https://vuejs.org/">Vue.js</a> as a front-end library.
        See the 
        <a href="https://github.com/TotallyInformation/node-red-contrib-uibuilder">node-red-contrib-uibuilder</a>
        README for details on how to use UIbuilder.
    </p>
    <h2>Dynamic Data (via Vue)</h2>

    <p>
        Messages: Received={{msgsReceived}}, Sent: {{msgsSent}}
    </p><p>
        Control Messages Received: {{msgsControl}}
    </p>

    <pre v-html="hLastRcvd"></pre>
    <pre v-html="hLastSent"></pre>

    <h2>Simple input using Vue</h2>
    <p><button v-on:click="increment">Increment</button> Click Counter: {{counterBtn}}. Click on the button to increment, it sends the data back to Node-RED as well.</p>
        
  </div>
</template>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
body {font-family: sans-serif;}
div, p, code, pre { margin:0.3em; padding: 0.3em;}
pre .string { color: green; }
.number { color: darkorange; }
.boolean { color: blue; }
.null { color: magenta; }
.key { color: red; }
</style>

```

**Comment by TotallyInformation** [comment]
Hmm, OK. I think this comes from the hoops we had to jump through originally in order to be able to isolate everything to a single object and allow it both to be loaded simply with no build phase **and** allow it to be used with webpack.

Unfortunately, I'm a bit lacking in knowledge at this level so I went with some code that was submitted in a PR.

I'd be happy if you were able to un/re-wrap this so it works in all cases. But the basic idea remains that it must "just work" with no need for a build and it must work if you do need to run a build.

**Comment by TotallyInformation** [comment]
@devleaks, having looked at this again, I can see that the problem lies with the way you are building your project.

I would need to understand more about how you are doing that. The `uibuilderfe.js` file should work just fine with something like webpack. I've not yet had a need to do that with my own projects and it is on the backlog of things to further develop.

In the meantime, you can either just live with loading the `min` version of `uibuilderfe.min.js` that I provide or you can pick apart the code and create an alternative version. If you do pick it apart, I'd love to see the result.

I've done an extensive home dashboard website now, using uibuilder and VueJS and it is working really well. But I'm still running Vue in dev mode - it is plenty fast enough for me and, as yet, I've no need to build separate components - I'm sure I will get to that at some point.

Leaving this open for a while longer in case you or someone else can help further.

**Comment by TotallyInformation** [comment]
OK, I've finally got round to testing this fully.

You have an error in your code. You are trying to load `./uibuilderfe.js` which may not exist.

I have a fully working example now with webpack v4.x loading a template from a .vue file with everything build and deployed to the `dist` folder in the `uibuilder/<url>` folder. I will create a new WIKI entry explaining the steps.

I cannot reproduce your error. If you try the basic example I'm about to put in the WIKI and still get the error then please feel free to reopen this issue. Otherwise, please raise a new issue.

https://github.com/TotallyInformation/node-red-contrib-uibuilder/wiki/Using-VueJS-with-Webpack

Thanks.
